Add DEMetropolisZ stepper

michaelosthege opened the pull request:
The DE-MCMC-Z is a variant of DE-MCMC that uses its own trace/history to make new proposals instead of using a population of chains.

Compared to `DEMetropolis` this has several advantages:
+ no need to step in parallel
+ no between-process communication when running with `cores > 1`
+ the "population" grows by every iteration -> works with `chains=1` already

In the current implementation the history is kept in the instance of the stepper even though the same information is already present in the trace. But memory is cheap compared to inter-process communication & having to fiddle with `sampling.py`,

I'm working on a benchmarking/comparison notebook and will also do unit tests..

michaelosthege committed:
implement DEMetropolisZ

junpenglao commented on pymc3/step_methods/metropolis.py:
```diff
@@ -616,6 +616,145 @@ def competence(var, has_grad):
+
+    def __init__(self, vars=None, S=None, proposal_dist=None, lamb=None, scaling=0.001,
+                 tune=None, tune_interval=100, model=None, mode=None, **kwargs):
+
```
need an Experimental warning.

junpenglao commented on pymc3/step_methods/metropolis.py:
```diff
@@ -616,6 +616,145 @@ def competence(var, has_grad):
+            # Reset counter
+            self.steps_until_tune = self.tune_interval
+            self.accepted = 0
+
```
Should `self._it` and `self._history` also be reset after tuning?

michaelosthege replied:
I'm not sure yet. My first tests show that DE-MCMC-Z has a slower initial convergence, but benefits greatly from tuning lambda. I'll run tests to answer your question.
Instead of `self._it` I can also do just `len(self._history)`.

michaelosthege replied:
Resetting the history after tuning is bad. Because each chain is on its own, its equivalent to dropping the entire history.
![image](https://user-images.githubusercontent.com/5894642/72918093-5d905800-3d45-11ea-855c-94bbbd3010df.png)

Maybe if one keeps the most recent 10%, but unless the chain started far away from the typical set even that should make it worse.

junpenglao replied:
Thanks, I guess my initial feeling is that it is getting samples from tuning stage which is usually not stationary. But it is true that we are not "just" accepting a history.

michaelosthege replied:
Did some further testing. It looks like a `tune_drop_fraction=0.9` does indeed improve efficiency when convergence to the typical set is slow.

![DEMetropolisZ_tuning_efficiency](https://user-images.githubusercontent.com/5894642/73015734-dca48f80-3e1c-11ea-956c-bea8eae77f6b.png)
(error bars indicate the standard error of the mean)

![DEMetropolisZ_tuning_autocorrelation](https://user-images.githubusercontent.com/5894642/73015740-df9f8000-3e1c-11ea-9b74-ed504cd8ec45.png)
![DEMetropolisZ_tuning_acceptance](https://user-images.githubusercontent.com/5894642/73015743-e3330700-3e1c-11ea-9e20-8377698ed25c.png)

junpenglao commented on pymc3/step_methods/metropolis.py:
```diff
@@ -616,6 +616,145 @@ def competence(var, has_grad):
+        accept = self.delta_logp(q, q0)
+        q_new, accepted = metrop_select(accept, q, q0)
+        self.accepted += accepted
+        self._history.append(q_new)
```
Is there any concern for memory for large num_step? It would be great if we can clear out `self._history` if user are calling from `pm.sample`

michaelosthege replied:
Initially I wanted to link it to the `BaseTrace`, but that's really difficult for many reasons.

I'd prefer to initialize an array instead of appending to a list, but the stepper doesn't know the number of draws ahead of time.

Also, re-using an instance of the step method is dangerous and I did not yet check if that happens in sequential singleprocess sampling (`_sample_many`).

aloctavodia commented on pymc3/step_methods/metropolis.py:
```diff
@@ -616,6 +616,145 @@ def competence(var, has_grad):
+        if it > 1:
+            # differential evolution proposal
+            # select two other chains
+            iz1 = np.random.randint(it)
```
what about
`iz1, iz2 = np.random.choice(range(it+1), size=2, replace=False)`

michaelosthege replied:
I benchmarked it (`for it in range(10000)`)with `np.arange` instead of `range` and it was waay slower than this (granted) ugly construct with the `while`

aloctavodia replied:
yeah random.choice is slow. Well, it's not that ugly :-)

junpenglao commented on pymc3/step_methods/metropolis.py:
```diff
@@ -616,6 +616,145 @@ def competence(var, has_grad):
         return Competence.COMPATIBLE
 
 
+class DEMetropolisZ(ArrayStepShared):
```
Could we use a more informative name?

michaelosthege replied:
Suggestions?
But in its behavior it's similar to `DEMetropolis` in many ways. For both algorithms the user would be well-advised to read the paper anyway.

junpenglao replied:
`DEHistoryMetropolis`?

aloctavodia replied:
The name used in the paper is DE-MCz. Given that we already have a class named DEMetropolis, DEMetropolisZ sounds like a good compromise to me.

junpenglao replied:
Cool, let's go with DEMetropolisZ then.

michaelosthege committed:
use len instead of attribute

michaelosthege committed:
introduce tune_drop_fraction
+ improves efficiency in situations with slow initial tuning convergence (see PR #3784)

michaelosthege committed:
reset tuned settings in sequential singleprocess sampling

michaelosthege committed:
tune DEMetropolisZ lambda by default

michaelosthege committed:
use new empty list to reset history

michaelosthege committed:
fix dual-use conflict of self.tune attribute

michaelosthege commented:
Here are some plots to better understand the strenghts and weaknesses of `DEMetropolisZ`:

### Tuning
These tests are done with `D=10`, `N_tune=10000`, `N_draws=10000`, `tune='lambda' ` and this model:

```
def get_mvnormal_model(D:int) -> pymc3.Model:
    true_mu = numpy.zeros(D)
    true_cov = numpy.eye(D)
    true_cov[:5,:5] = numpy.array([
        [1, 0.5, 0, 0, 0],
        [0.5, 2, 2, 0, 0],
        [0, 2, 3, 0, 0],
        [0, 0, 0, 4, 4],
        [0, 0, 0, 4, 5],
    ])

    with pymc3.Model() as pmodel:
        x = pymc3.MvNormal('x', mu=true_mu, cov=true_cov, shape=(D,))
    
    true_samples = x.random(size=1000)
    truth_id = arviz.data.convert_to_inference_data(true_samples[numpy.newaxis,:], group='random')
    return pmodel, truth_id
```

The `tune_drop_fraction` parameter helps to increase effective sample sizes by discarding parts of the tuning history.

![image](https://user-images.githubusercontent.com/5894642/73244137-7ff0fe00-41a9-11ea-8949-23241187b3c0.png)

When the entire tuning history is removed, the autocorrelation goes up:

![image](https://user-images.githubusercontent.com/5894642/73244148-897a6600-41a9-11ea-929b-d34ccb4f736a.png)

Similarly, the acceptance rate jumps up:

![image](https://user-images.githubusercontent.com/5894642/73244210-af076f80-41a9-11ea-854c-173b51bd2a13.png)

Tuning `lambda` helps the initial convergence (same run as above). The chains were started at `[5.0]*N_chains` to put them outside of the typical set.

![image](https://user-images.githubusercontent.com/5894642/73244243-c7778a00-41a9-11ea-8157-ef1d8fc10a7b.png)

### Efficiency compared to `DEMetropolis`
Compared to `DEMetropolis`, the `DEMetropolisZ` achieves the same effective sample sizes independent of the number of chains. At `N_chains=2*D`, `DEMetropolis` almost catches up:

![image](https://user-images.githubusercontent.com/5894642/73244603-9ba8d400-41aa-11ea-9013-8ea9b2ba3299.png)

I ran all experiments with `cores=1` and there was no significant runtime difference.

junpenglao commented:
Thanks @michaelosthege - I think comments like https://github.com/pymc-devs/pymc3/pull/3784#issuecomment-579122223 would be best in our wiki, so we can link it in the docstring if necessary.

michaelosthege commented:
> Thanks @michaelosthege - I think comments like #3784 (comment) would be best in our wiki, so we can link it in the docstring if necessary.

I have two or three notebooks for running these experiments that I would like to contribute in some way.
Maybe the "Deep Dive" section in the [Tutorials](https://docs.pymc.io/nb_tutorials/index.html) ?

Also I should probably move the "pymc3.ode: Shapes and benchmarking" [Example](https://docs.pymc.io/nb_examples/index.html) over to that section too.

michaelosthege committed:
add warning and initialize self.tune attribute

michaelosthege committed:
add tests for DEMetropolisZ

michaelosthege committed:
Incorporate recent updates from master

michaelosthege committed:
add tests to hit more of the diff

michaelosthege committed:
Merge branch 'demcmc-z' of https://github.com/michaelosthege/pymc3 into demcmc-z

michaelosthege committed:
add and move notebooks, update release notes

canyon289 commented on pymc3/step_methods/metropolis.py:
```diff
@@ -630,6 +631,176 @@ def competence(var, has_grad):
+            lamb = 2.38 / np.sqrt(2 * model.ndim)
+        self.lamb = float(lamb)
+        if not tune in {None, 'scaling', 'lambda'}:
+            raise ValueError('The parameter "tune" must be one of {None, scaling, lambda}')
```
This is very picky; Would be easier to read if written `if  tune not in {None, 'scaling', 'lambda'}:` because then it reads like english instead of like a program. Not implying that a change is required

canyon289 commented on pymc3/step_methods/metropolis.py:
```diff
@@ -630,6 +631,176 @@ def competence(var, has_grad):
+
+        self.scaling = np.atleast_1d(scaling).astype('d')
+        if lamb is None:
+            lamb = 2.38 / np.sqrt(2 * model.ndim)
```
What's 2.38? Can a comment be left for the magic number?

aloctavodia replied:
This comes from theoretical studies about the optimal covariance matrix and optimal acceptance rate for MCMC methods. This papers collects many of these findings https://people.eecs.berkeley.edu/~jordan/sail/readings/andrieu-thoms.pdf another magic number is 0.234, the theoretical optimal acceptance rate for a random walk MC as the number of dimensions goes to infinity, empirically infinite happens around at 5 :)

canyon289 commented on pymc3/step_methods/metropolis.py:
```diff
@@ -630,6 +631,176 @@ def competence(var, has_grad):
+        return q_new, [stats]
+
+    def stop_tuning(self):
+        # remove the first x% of the tuning phase so future steps are not
```
Can this be made a standard triple quotes comment? That way it shows up with `help()` and doc discovery tools, such as the ones in ides

canyon289 commented on pymc3/tests/test_step.py:
```diff
@@ -778,6 +779,137 @@ def test_parallelized_chains_are_random(self):
+            assert trace.get_sampler_stats('scaling', chains=c)[0] == 0.002
+            assert trace.get_sampler_stats('scaling', chains=c)[-1] != 0.002
+            assert set(trace.get_sampler_stats('tune', chains=c)) == {True, False}
+        pass
```
What is this pass doing?

michaelosthege replied:
I started appending `pass` to tests to mark the end of the case (as a protection against accidental deletion of lines).

canyon289 replied:
Ah ok. Doesn't hurt anything, just the first time I've seen it!

canyon289 commented on pymc3/tests/test_step.py:
```diff
@@ -778,6 +779,137 @@ def test_parallelized_chains_are_random(self):
+    def test_invalid_tune(self):
+        with Model() as pmodel:
+            Normal('n', 0, 2, shape=(3,))
+            with pytest.raises(ValueError):
```
It's worth parameterizing this so in the vent multiple tests fail its easier to figure out which ones are failing. Right now earlier test failures will hide the failures of the subsequent ones

canyon289 commented on pymc3/tests/test_step.py:
```diff
@@ -778,6 +779,137 @@ def test_parallelized_chains_are_random(self):
+        with Model() as pmodel:
+            n = Normal('n', 0, 2, shape=(3,))
+            b = Binomial('b', n=2, p=0.3)
+        assert DEMetropolisZ.competence(n, has_grad=True) == 1
```
Same comment here about parameterization

It's worth parameterizing this so in the vent multiple tests fail its easier to figure out which ones are failing. Right now earlier test failures will hide the failures of the subsequent ones

canyon289 commented on pymc3/tests/test_step.py:
```diff
@@ -778,6 +779,137 @@ def test_parallelized_chains_are_random(self):
+                discard_tuned_samples=False
+            )
+            assert len(trace) == 500
+            assert len(step._history) == (300 - 300 * 0.85) + 200
```
Being picky here. This test would be easier to read if written like this

```
tune =300
tune_drop_fraction=.85
draws=200

            step = DEMetropolisZ(tune_drop_fraction=tune_drop_fraction)
            trace = sample(
                tune=tune,
                draws=draws,
                step=step,
                cores=1,
                chains=1,
                discard_tuned_samples=False
            )

assert len(step._history) == (tune - tune * tune_drop_fraction) + draws
```

That way there's less mental steps to correlate where the step_history expectation is coming from. Same for trace

canyon289 commented:
Very well written! Thanks for adding to PyMC3! Left some code style comments. Great work!!

twiecki commented on RELEASE-NOTES.md:
```diff
@@ -5,6 +5,8 @@
 ### New features
 - use [fastprogress](https://github.com/fastai/fastprogress) instead of tqdm [#3693](https://github.com/pymc-devs/pymc3/pull/3693)
 - `DEMetropolis` can now tune both `lambda` and `scaling` parameters, but by default neither of them are tuned. See [#3743](https://github.com/pymc-devs/pymc3/pull/3743) for more info.
+- `DEMetropolisZ`, an improved variant of `DEMetropolisZ` brings better parallelization and high efficiency with few chains with a slower initial convergence. This implementation is experimental. See [#3784](https://github.com/pymc-devs/pymc3/pull/3784) for more info.
```
```suggestion
- `DEMetropolisZ`, an improved variant of `DEMetropolis` brings better parallelization and high efficiency with few chains with a slower initial convergence. This implementation is experimental. See [#3784](https://github.com/pymc-devs/pymc3/pull/3784) for more info.
```

twiecki commented on RELEASE-NOTES.md:
```diff
@@ -5,6 +5,8 @@
 - use [fastprogress](https://github.com/fastai/fastprogress) instead of tqdm [#3693](https://github.com/pymc-devs/pymc3/pull/3693)
 - `DEMetropolis` can now tune both `lambda` and `scaling` parameters, but by default neither of them are tuned. See [#3743](https://github.com/pymc-devs/pymc3/pull/3743) for more info.
+- `DEMetropolisZ`, an improved variant of `DEMetropolisZ` brings better parallelization and high efficiency with few chains with a slower initial convergence. This implementation is experimental. See [#3784](https://github.com/pymc-devs/pymc3/pull/3784) for more info.
+- Notebooks that give insight into `DEMetropolis`, `DEMetropolisZ` and the `DifferentialEquation` interface are now located in the [Tutorials/Deep Dive]() section.
```
Need link.

michaelosthege committed:
address review feedback

aloctavodia commented on docs/source/notebooks/DEMetropolisZ_tune_drop_fraction.ipynb:
```diff
@@ -0,0 +1,1795 @@
+    "import time\n",
+    "import pymc3\n",
+    "import arviz\n",
+    "import theano.tensor as tt"
```
I think this import is not used. Also could we follow the convention of using alias when importing the libraries like `import numpy as np`, I think this is the style we follow in most notebooks.

michaelosthege replied:
I can do that. However I really don't like this "convention" of aliasing a 5-letter package into a 2-letter alias. I've seen many beginners being super confused by this.

aloctavodia commented on docs/source/notebooks/DEMetropolisZ_EfficiencyComparison.ipynb:
```diff
@@ -0,0 +1,2066 @@
+    "import fastprogress\n",
+    "import pymc3\n",
+    "import arviz\n",
+    "import theano.tensor as tt"
```
same as my other comment

aloctavodia commented on docs/source/notebooks/DEMetropolisZ_EfficiencyComparison.ipynb:
```diff
@@ -0,0 +1,2066 @@
+   "outputs": [
+    {
+     "data": {
+      "image/png": "iVBORw0KGgoAAAANSUhEUgAABXwAAALUCAYAAAC4mfpWAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAVhwAAFYcBshnuugAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nOzdeZhcZ30n+u/pRXtLlrC8L5KNvAjbYIY1EwxJIDAzCUmAmwQCgyEwkzyBDJkJMPfCZUluZiAZyABzkztjIAswhISQPZAwGQiEsCaATYSNjW15wfKCZHVrV3ef+8fb7a6qru6ubnV3dbU+n+c5T52uOu85b7Utqepbv/q9VV3XAQAAAACg9/V ...
```
Shouldn't be "DEMetropolisZ can achieve the same effective sample sizes with less chains."

aloctavodia reviewed (approved):
minor details

michaelosthege committed:
apply review suggestions

twiecki commented on RELEASE-NOTES.md:
```diff
@@ -5,6 +5,8 @@
 ### New features
 - use [fastprogress](https://github.com/fastai/fastprogress) instead of tqdm [#3693](https://github.com/pymc-devs/pymc3/pull/3693)
 - `DEMetropolis` can now tune both `lambda` and `scaling` parameters, but by default neither of them are tuned. See [#3743](https://github.com/pymc-devs/pymc3/pull/3743) for more info.
+- `DEMetropolisZ`, an improved variant of `DEMetropolis` brings better parallelization and high efficiency with few chains with a slower initial convergence. This implementation is experimental. See [#3784](https://github.com/pymc-devs/pymc3/pull/3784) for more info.
```
```suggestion
- `DEMetropolisZ`, an improved variant of `DEMetropolis` brings better parallelization and higher efficiency with fewer chains with a slower initial convergence. This implementation is experimental. See [#3784](https://github.com/pymc-devs/pymc3/pull/3784) for more info.
```

michaelosthege committed:
more accurate wording

Co-Authored-By: Thomas Wiecki <thomas.wiecki@gmail.com>
